docs/howtos/lr_schedule.rst

```diff
@@ -1,138 +1,230 @@
+        k: np.mean([metrics[k] for metrics in batch_metrics])
+        for k in batch_metrics[0]}
+  
+    logging.info('train epoch: %d, loss: %.4f, accuracy: %.2f', epoch,
+                 epoch_metrics['loss'], epoch_metrics['accuracy'] * 100)
+  
+    return state, epoch_metrics
   ---
-  def train_epoch(optimizer, train_ds, batch_size, epoch, rng):
-    """Train for a single epoch."""
+  def train_epoch(state, train_ds, batch_size, epoch, learning_rate_fn, rng): #!
+    """Trains for a single epoch."""
     train_ds_size = len(train_ds['image'])
     steps_per_epoch = train_ds_size // batch_size
-    # 4 cycles per epoch #!
-    learning_rate_fn = create_triangular_schedule( #!
-      3e-3, 3e-2, steps_per_epoch // 4) #!
     perms = jax.random.permutation(rng, len(train_ds['image']))
     perms = perms[:steps_per_epoch * batch_size]
     perms = perms.reshape((steps_per_epoch, batch_size))
     batch_metrics = []
     for perm in perms:
       batch = {k: v[perm, ...] for k, v in train_ds.items()}
-      optimizer, metrics = train_step(optimizer, batch, learning_rate_fn) #!
+      state, metrics = train_step(state, batch, learning_rate_fn) #!
       batch_metrics.append(metrics)
-
+  
     # compute mean of metrics across each batch in epoch.
     batch_metrics = jax.device_get(batch_metrics)
     epoch_metrics = {
         k: np.mean([metrics[k] for metrics in batch_metrics])
         for k in batch_metrics[0]}
-
+  
     logging.info('train epoch: %d, loss: %.4f, accuracy: %.2f', epoch,
-                epoch_metrics['loss'], epoch_metrics['accuracy'] * 100)
+                 epoch_metrics['loss'], epoch_metrics['accuracy'] * 100)
+  
+    return state, epoch_metrics
+
+
+And the ``create_train_state`` function:
+
+
+.. codediff::
+  :title_left: Default learning rate
+  :title_right: Learning rate schedule
+
+  def create_train_state(rng, config):
+    """Creates initial `TrainState`."""
+    cnn = CNN()
+    params = cnn.init(rng, jnp.ones([1, 28, 28, 1]))['params']
+    tx = optax.sgd(config.learning_rate, config.momentum)
+    return train_state.TrainState.create(
+        apply_fn=cnn.apply, params=params, tx=tx)
+  ---
+  def create_train_state(rng, config, learning_rate_fn): #!
+    """Creates initial `TrainState`."""
+    cnn = CNN()
+    params = cnn.init(rng, jnp.ones([1, 28, 28, 1]))['params']
+    tx = optax.sgd(learning_rate_fn, config.momentum) #!
+    return train_state.TrainState.create(
+        apply_fn=cnn.apply, params=params, tx=tx)
+
+
+
+.. testcode::
+
+  config = get_config()
+  
+  train_ds_size = config.train_ds_size
+  steps_per_epoch = train_ds_size // config.batch_size
+  learning_rate_fn = create_learning_rate_fn(config, config.learning_rate, steps_per_epoch)
+  
+  rng = jax.random.PRNGKey(0)
+  state = create_train_state(rng, config, learning_rate_fn)
+  rng, _ = jax.random.split(rng)
+
+  batch = get_dummy_batch(config.batch_size)
+  state, metrics = train_step(state, batch, learning_rate_fn)
```

Comment:
Why not set `config.train_ds_size=128` and then call `train_epoch()` instead? That shouldn't take significantly longer and would test all your code.